tests/unit/s2n_cipher_suite_match_test.c

```diff
@@ -781,6 +781,295 @@ int main(int argc, char **argv)
+                        &s2n_tls13_aes_128_gcm_sha256,                  /* 0x13,0x01 */
+                        &s2n_tls13_aes_256_gcm_sha384,                  /* 0x13,0x02 */
+                        &s2n_tls13_chacha20_poly1305_sha256,            /* 0x13,0x03 */
+                    };
+
+                    const struct s2n_cipher_preferences cipher_preference_with_chacha20_and_no_boosting = {
+                        .count = s2n_array_len(cipher_suites_with_chacha20_last),
+                        .suites = cipher_suites_with_chacha20_last,
+                        .allow_chacha20_boosting = false
+                    };
+
+                    const struct s2n_security_policy test_chacha20_and_not_boosted = {
+                        .minimum_protocol_version = S2N_SSLv3,
+                        .cipher_preferences = &cipher_preference_with_chacha20_and_no_boosting,
+                        .kem_preferences = &kem_preferences_null,
+                        .signature_preferences = &s2n_signature_preferences_20201021,
+                        .ecc_preferences = &s2n_ecc_preferences_test_all,
+                    };
+
+                    uint8_t chacha20_boosted_wire[] = {
+                        TLS_CHACHA20_POLY1305_SHA256,
+                        TLS_AES_256_GCM_SHA384,
+                        TLS_AES_128_GCM_SHA256,
+                    };
+
+                    conn->security_policy_override = &test_chacha20_and_not_boosted;
+                    conn->kex_params.server_ecc_evp_params.negotiated_curve = s2n_all_supported_curves_list[0];
+                    conn->client_protocol_version = S2N_TLS13;
+                    conn->actual_protocol_version = S2N_TLS13;
+                    conn->server_protocol_version = S2N_TLS13;
+
+                    uint8_t count = sizeof(chacha20_boosted_wire) / S2N_TLS_CIPHER_SUITE_LEN;
+                    EXPECT_SUCCESS(s2n_set_cipher_as_tls_server(conn, chacha20_boosted_wire, count));
+                    /* Server has chacha20 boosting disabled so the server negotiates based on server preference */
+                    EXPECT_EQUAL(conn->secure->cipher_suite, &s2n_tls13_aes_128_gcm_sha256);
+
+                    uint8_t chacha20_only_wire[] = {
+                        TLS_CHACHA20_POLY1305_SHA256,
+                    };
+
+                    count = sizeof(chacha20_only_wire) / S2N_TLS_CIPHER_SUITE_LEN;
+
+                    EXPECT_SUCCESS(s2n_set_cipher_as_tls_server(conn, chacha20_only_wire, count));
+                    /* Server has chacha20 boosting disabled but we can still negotiate it if it's the only option */
+                    EXPECT_EQUAL(conn->secure->cipher_suite, &s2n_tls13_chacha20_poly1305_sha256);
+
+                    EXPECT_SUCCESS(s2n_connection_wipe(conn));
+                }
+
+                /* Chacha20 boosting support is enabled server-side with multiple potential chacha20 cipher suites */
+                {
```

Comment:
I don't see the case where the highest priority chacha20 suite on the server's list isn't available, so the server needs to choose the next highest chacha20 option.